File: packages/core/wurstbrot_core/solver.py
from __future__ import annotations

from dataclasses import replace
from heapq import heappop, heappush
from itertools import count

from .database import DatabaseError, VehicleDatabase
from .economy import ALLOWED_SL_DISCOUNTS, apply_discount, ge_for_remaining_rp
from .models import (
    PlayerProgress,
    RankRequirement,
    SolveOptions,
    SolveResult,
    Vehicle,
    VehicleCostLine,
    VehicleProgress,
)


class SolveError(ValueError):
    pass


def _nonnegative_int(value: object) -> bool:
    return isinstance(value, int) and not isinstance(value, bool) and value >= 0


class ResearchSolver:
    def __init__(self, database: VehicleDatabase) -> None:
        self.db = database

    def solve(
        self,
        *,
        target_vehicle_id: str,
        start_vehicle_id: str | None = None,
        progress: PlayerProgress | None = None,
        options: SolveOptions | None = None,
    ) -> SolveResult:
        progress = progress or PlayerProgress()
        options = options or SolveOptions()

        self._validate_input_contract(progress, options)

        target = self.db.get(target_vehicle_id)
        start = self.db.get(start_vehicle_id) if start_vehicle_id else None

        if start and (
            start.country_id != target.country_id
            or start.branch_id != target.branch_id
        ):
            raise SolveError("Start und Ziel müssen im selben Forschungsbaum liegen.")

        if target.hidden_research and not options.include_hidden_legacy:
            raise SolveError(
                f"{target.name} ist ein ausgeblendetes Altbestandsfahrzeug."
            )

        direct_path = list(self.db.closure(target.id))

        # Vehicles marked owned imply their mandatory predecessors. Vehicle A may
        # be in another line of the same tree: owning A proves its own prerequisite
        # chain and all lower rank gates, but not the target line's predecessors.
        owned = self._expanded_owned(progress, target.country_id, target.branch_id)
        if start:
            start_closure = set(self.db.closure(start.id))
            if options.include_start_vehicle:
                owned.update(start_closure - {start.id})
            else:
                owned.update(start_closure)

        direct_required = [
            vehicle_id for vehicle_id in direct_path if vehicle_id not in owned
        ]
        owned.update(
            vehicle.id
            for vehicle in self.db.tree_vehicles(target.country_id, target.branch_id)
            if vehicle.reserve
        )

        required: set[str] = set(direct_required)
        reasons: dict[str, str] = {
            vehicle_id: "direct_path" for vehicle_id in direct_required
        }

        rank_requirements: list[RankRequirement] = []
        # If the player can already research/own vehicle A, all rank gates below
        # A's rank have necessarily been passed in-game. Rechecking them would
        # invent obsolete costs. A's own rank still matters when B is in a later rank.
        first_rank_to_check = start.rank if start else 1
        for rank in range(first_rank_to_check, target.rank):
            needed = self.db.rank_requirement(
                target.country_id, target.branch_id, rank
            )
            if needed <= 0:
                continue

            before_set = owned | required
            available_before = self._count_rank(before_set, rank)
            added: set[str] = set()

            if available_before < needed:
                added = self._find_minimum_rank_additions(
                    base=before_set,
                    country_id=target.country_id,
                    branch_id=target.branch_id,
                    rank=rank,
                    required_count=needed,
                    progress=progress,
                    options=options,
                    allow_req_unlock=bool(start) or bool(owned),
                )
                required.update(added)
                for vehicle_id in added:
                    reasons.setdefault(vehicle_id, "rank_unlock")

            available_after = self._count_rank(owned | required, rank)
            if available_after < needed:
                raise SolveError(
                    f"Rang {rank + 1} kann nicht geöffnet werden: "
                    f"{available_after}/{needed} Fahrzeuge."
                )

            rank_requirements.append(
                RankRequirement(
                    rank=rank,
                    required=needed,
                    available_before=available_before,
                    available_after=available_after,
                    added_vehicle_ids=tuple(
                        sorted(
                            (
                                vehicle_id
                                for vehicle_id in added
                                if self.db.get(vehicle_id).rank == rank
                            ),
                            key=self._vehicle_sort_key,
                        )
                    ),
                )
            )

        if start and options.include_start_vehicle:
            reasons[start.id] = "start_vehicle"
            required.add(start.id)

        ordered_required = sorted(required, key=self._vehicle_sort_key)
        lines: list[VehicleCostLine] = []
        warnings: list[str] = []

        for vehicle_id in ordered_required:
            vehicle = self.db.get(vehicle_id)
            vehicle_progress = progress.for_vehicle(vehicle_id)
            already_owned = vehicle_progress.owned or vehicle_id in owned

            researched_rp = vehicle_progress.researched_rp
            remaining_rp = 0 if already_owned else max(vehicle.rp - researched_rp, 0)
            ge = ge_for_remaining_rp(remaining_rp, self.db.rp_per_ge)
            sl = 0 if already_owned else apply_discount(
                vehicle.sl, options.sl_discount_percent
            )

            if vehicle.req_unlock:
                warnings.append(
                    f"{vehicle.name}: zusätzliche Freischaltung {vehicle.req_unlock}"
                )
            if vehicle.hidden_research:
                warnings.append(f"{vehicle.name}: Altbestandsfahrzeug")

            lines.append(
                VehicleCostLine(
                    vehicle_id=vehicle.id,
                    name=vehicle.name,
                    reason=reasons.get(vehicle_id, "rank_unlock"),
                    total_rp=vehicle.rp,
                    researched_rp=researched_rp,
                    remaining_rp=remaining_rp,
                    ge=ge,
                    sl=sl,
                    already_owned=already_owned,
                )
            )

        total_rp = sum(line.remaining_rp for line in lines)
        total_ge = sum(line.ge for line in lines)
        total_sl = sum(line.sl for line in lines)
        convertible_shortfall = (
            max(total_rp - progress.convertible_rp, 0)
            if progress.convertible_rp is not None
            else 0
        )

        return SolveResult(
            start_vehicle_id=start.id if start else None,
            target_vehicle_id=target.id,
            vehicle_lines=tuple(lines),
            rank_requirements=tuple(rank_requirements),
            required_vehicle_ids=tuple(line.vehicle_id for line in lines),
            total_rp=total_rp,
            total_ge_before_owned=total_ge,
            total_ge_after_owned=max(total_ge - progress.owned_ge, 0),
            total_sl=total_sl,
            convertible_rp_shortfall=convertible_shortfall,
            warnings=tuple(dict.fromkeys(warnings)),
        )

    def _expanded_owned(
        self,
        progress: PlayerProgress,
        country_id: str,
        branch_id: str,
    ) -> set[str]:
        result: set[str] = set()
        for vehicle_id, state in progress.vehicles.items():
            if not state.owned or vehicle_id not in self.db.vehicles:
                continue
            vehicle = self.db.get(vehicle_id)
            if vehicle.country_id == country_id and vehicle.branch_id == branch_id:
                result.update(self.db.closure(vehicle_id))
        return result

    def _count_rank(self, vehicle_ids: set[str], rank: int) -> int:
        return sum(
            1
            for vehicle_id in vehicle_ids
            if vehicle_id in self.db.vehicles
            and self.db.get(vehicle_id).rank == rank
        )

    def _vehicle_sort_key(self, vehicle_id: str) -> tuple:
        vehicle = self.db.get(vehicle_id)
        return (
            vehicle.rank,
            vehicle.column,
            vehicle.order,
            vehicle.id,
        )

    def _candidate_cost(
        self,
        vehicle_ids: set[str],
        base: set[str],
        progress: PlayerProgress,
        options: SolveOptions,
    ) -> tuple[int, int, int, tuple[str, ...]]:
        new_ids = vehicle_ids - base
        rp = 0
        ge = 0
        sl = 0

        for vehicle_id in new_ids:
            vehicle = self.db.get(vehicle_id)
            state = progress.for_vehicle(vehicle_id)
            if state.owned or vehicle.reserve:
                continue
            researched_rp = state.researched_rp
            remaining_rp = max(vehicle.rp - researched_rp, 0)
            rp += remaining_rp
            ge += ge_for_remaining_rp(remaining_rp, self.db.rp_per_ge)
            sl += apply_discount(vehicle.sl, options.sl_discount_percent)

        if options.optimize_for == "sl":
            primary = sl
        elif options.optimize_for == "vehicles":
            primary = len(new_ids)
        elif options.optimize_for == "rp":
            primary = rp
        else:
            primary = ge

        # Deterministic tie breakers.
        return (primary, ge, sl, tuple(sorted(new_ids)))

    def _validate_input_contract(
        self,
        progress: PlayerProgress,
        options: SolveOptions,
    ) -> None:
        """Enforce the evidence-backed version-1.0 user input contract."""
        if not isinstance(progress, PlayerProgress):
            raise SolveError("PlayerProgress ist ungültig.")
        if not _nonnegative_int(progress.owned_ge):
            raise SolveError("owned_ge muss eine nicht-negative Ganzzahl sein.")
        if progress.convertible_rp is not None and not _nonnegative_int(
            progress.convertible_rp
        ):
            raise SolveError(
                "convertible_rp muss null oder eine nicht-negative Ganzzahl sein."
            )
        if not isinstance(progress.fulfilled_unlocks, (set, frozenset)) or any(
            not isinstance(item, str) or not item
            for item in progress.fulfilled_unlocks
        ):
            raise SolveError(
                "fulfilled_unlocks darf nur nichtleere Zeichenketten enthalten."
            )
        if not isinstance(progress.vehicles, dict):
            raise SolveError("progress.vehicles muss eine Zuordnung sein.")

        for vehicle_id, state in sorted(
            progress.vehicles.items(), key=lambda item: str(item[0])
        ):
            if vehicle_id not in self.db.vehicles:
                raise SolveError(
                    f"PlayerProgress enthält ein unbekanntes Fahrzeug: {vehicle_id}"
                )
            if not isinstance(state, VehicleProgress):
                raise SolveError(
                    f"Ungültiger Fortschrittsstatus für {vehicle_id}."
                )
            vehicle = self.db.get(vehicle_id)
            if not _nonnegative_int(state.researched_rp):
                raise SolveError(
                    f"researched_rp für {vehicle_id} muss eine nicht-negative "
                    "Ganzzahl sein."
                )
            if state.researched_rp > vehicle.rp:
                raise SolveError(
                    f"researched_rp für {vehicle_id} überschreitet die Fahrzeug-RP."
                )
            if not isinstance(state.researched, bool) or not isinstance(
                state.purchased, bool
            ):
                raise SolveError(
                    f"researched und purchased für {vehicle_id} müssen boolesch sein."
                )
            if state.purchased and not state.researched:
                raise SolveError(
                    f"Ein gekauftes Fahrzeug muss als erforscht markiert sein: {vehicle_id}"
                )
            if state.researched and state.researched_rp != vehicle.rp:
                raise SolveError(
                    f"researched=True für {vehicle_id} erfordert exakt "
                    f"{vehicle.rp} researched_rp."
                )

        if not isinstance(options, SolveOptions):
            raise SolveError("SolveOptions ist ungültig.")
        if not isinstance(options.optimize_for, str) or options.optimize_for not in {
            "ge",
            "rp",
            "sl",
            "vehicles",
        }:
            raise SolveError("optimize_for muss ge, rp, sl oder vehicles sein.")
        for field_name in (
            "include_start_vehicle",
            "include_hidden_legacy",
            "assume_external_unlocks",
        ):
            if not isinstance(getattr(options, field_name), bool):
                raise SolveError(f"{field_name} muss boolesch sein.")
        if (
            not isinstance(options.sl_discount_percent, int)
            or isinstance(options.sl_discount_percent, bool)
            or options.sl_discount_percent not in ALLOWED_SL_DISCOUNTS
        ):
            raise SolveError("SL-Rabatt muss 0, 30 oder 50 Prozent betragen.")

    def _find_minimum_rank_additions(
        self,
        *,
        base: set[str],
        country_id: str,
        branch_id: str,
        rank: int,
        required_count: int,
        progress: PlayerProgress,
        options: SolveOptions,
        allow_req_unlock: bool,
    ) -> set[str]:
        candidates = [
            vehicle
            for vehicle in self.db.tree_vehicles(country_id, branch_id)
            if vehicle.rank == rank
            and vehicle.id not in base
            and (options.include_hidden_legacy or not vehicle.hidden_research)
            and (allow_req_unlock or not vehicle.req_unlock)
        ]

        if not candidates:
            return set()

        # Uniform-cost search over unique prerequisite closures.
        counter = count()
        heap: list[tuple[tuple, int, frozenset[str]]] = []
        start_state = frozenset()
        heappush(
            heap,
            (self._candidate_cost(set(), base, progress, options), next(counter), start_state),
        )
        visited: set[frozenset[str]] = set()

        # A safety limit prevents pathological explosion while keeping normal
        # War Thunder trees exact for the small number of vehicles needed per rank.
        max_states = 75_000
        processed = 0

        while heap:
            _, _, state = heappop(heap)
            if state in visited:
                continue
            visited.add(state)
            processed += 1
            if processed > max_states:
                raise SolveError(
                    "Rangoptimierung hat das Sicherheitslimit erreicht."
                )

            combined = base | set(state)
            if self._count_rank(combined, rank) >= required_count:
                return set(state)

            for candidate in candidates:
                if candidate.id in combined:
                    continue
                closure = {
                    vehicle_id
                    for vehicle_id in self.db.closure(candidate.id)
                    if vehicle_id not in base
                    and (
                        options.include_hidden_legacy
                        or not self.db.get(vehicle_id).hidden_research
                    )
                }
                next_state = frozenset(set(state) | closure)
                if next_state in visited:
                    continue
                heappush(
                    heap,
                    (
                        self._candidate_cost(
                            set(next_state), base, progress, options
                        ),
                        next(counter),
                        next_state,
                    ),
                )

        return set()

File: apps/web/solver.mjs
const validatedDatabases = new WeakSet();

export function validateDatabase(raw) {
  if (raw && typeof raw === "object" && validatedDatabases.has(raw)) return raw;
  if (!raw || raw.schemaVersion !== 1) throw new Error("Nicht unterstützte Datenbank.");
  if (!Array.isArray(raw.vehicles) || !raw.vehicles.length) {
    throw new Error("Keine Fahrzeuge gefunden.");
  }
  const ids = new Set();
  for (const vehicle of raw.vehicles) {
    if (!vehicle.id || ids.has(vehicle.id)) {
      throw new Error(`Ungültige Fahrzeug-ID: ${vehicle.id || "leer"}`);
    }
    ids.add(vehicle.id);
  }
  for (const [id, predecessor] of Object.entries(raw.predecessors || {})) {
    if (ids.has(id) && predecessor && !ids.has(predecessor)) {
      throw new Error(`Unbekannter Vorgänger: ${predecessor}`);
    }
  }
  const rpPerGe = Number(raw.economy?.rpPerGE ?? 45);
  if (!Number.isInteger(rpPerGe) || rpPerGe <= 0) {
    throw new Error("rpPerGE muss eine positive Ganzzahl sein.");
  }
  for (const vehicle of raw.vehicles) {
    const seen = new Set();
    let current = vehicle.id;
    while (current) {
      if (seen.has(current)) throw new Error(`Zyklus im Forschungsweg bei ${current}`);
      if (!ids.has(current)) throw new Error(`Unbekanntes Fahrzeug: ${current}`);
      seen.add(current);
      current = (raw.predecessors || {})[current] || null;
    }
  }
  validatedDatabases.add(raw);
  return raw;
}

export function vehicleIndex(database) {
  return new Map(database.vehicles.map(vehicle => [vehicle.id, vehicle]));
}

export function closure(database, targetId) {
  const index = vehicleIndex(database);
  if (!index.has(targetId)) throw new Error(`Unbekanntes Fahrzeug: ${targetId}`);
  const result = [];
  const seen = new Set();
  let current = targetId;
  while (current) {
    if (seen.has(current)) throw new Error(`Zyklus im Forschungsweg bei ${current}`);
    if (!index.has(current)) throw new Error(`Unbekanntes Fahrzeug: ${current}`);
    seen.add(current);
    result.push(current);
    current = (database.predecessors || {})[current] || null;
  }
  return result.reverse();
}

function halfEvenDivide(numerator, denominator) {
  const quotient = Math.floor(numerator / denominator);
  const remainder = numerator % denominator;
  if (remainder * 2 < denominator) return quotient;
  if (remainder * 2 > denominator) return quotient + 1;
  return quotient % 2 === 0 ? quotient : quotient + 1;
}

function applyDiscount(value, percent) {
  if (!Number.isInteger(percent) || percent < 0 || percent > 100) {
    throw new Error("Rabatt muss zwischen 0 und 100 liegen.");
  }
  return halfEvenDivide(value * (100 - percent), 100);
}

function compareTuples(left, right) {
  for (let i = 0; i < Math.max(left.length, right.length); i += 1) {
    if (left[i] < right[i]) return -1;
    if (left[i] > right[i]) return 1;
  }
  return 0;
}

class MinHeap {
  constructor(compare) { this.items = []; this.compare = compare; }
  get size() { return this.items.length; }
  push(value) {
    this.items.push(value);
    for (let index = this.items.length - 1; index > 0;) {
      const parent = Math.floor((index - 1) / 2);
      if (this.compare(this.items[parent], this.items[index]) <= 0) break;
      [this.items[parent], this.items[index]] = [this.items[index], this.items[parent]];
      index = parent;
    }
  }
  pop() {
    const first = this.items[0];
    const last = this.items.pop();
    if (this.items.length && last) {
      this.items[0] = last;
      for (let index = 0;;) {
        const left = index * 2 + 1;
        const right = left + 1;
        let smallest = index;
        if (left < this.items.length && this.compare(this.items[left], this.items[smallest]) < 0) smallest = left;
        if (right < this.items.length && this.compare(this.items[right], this.items[smallest]) < 0) smallest = right;
        if (smallest === index) break;
        [this.items[index], this.items[smallest]] = [this.items[smallest], this.items[index]];
        index = smallest;
      }
    }
    return first;
  }
}

function stateFor(progress, id) {
  return progress.vehicles?.[id] || {};
}

function isOwned(state) {
  return Boolean(state.researched && state.purchased);
}

function sortKey(vehicle) {
  return [Number(vehicle.rank), Number(vehicle.column || 0), Number(vehicle.order || 0), vehicle.id];
}

function treeVehicles(database, countryId, branchId) {
  return database.vehicles
    .filter(vehicle => vehicle.countryId === countryId && vehicle.branchId === branchId)
    .sort((a, b) => compareTuples(sortKey(a), sortKey(b)));
}

function countRank(ids, rank, index) {
  let total = 0;
  for (const id of ids) if (index.get(id)?.rank === rank) total += 1;
  return total;
}

function candidateCost(ids, base, index, progress, options, rpPerGe) {
  const newIds = [...ids].filter(id => !base.has(id));
  let rp = 0; let ge = 0; let sl = 0;
  for (const id of newIds) {
    const vehicle = index.get(id);
    const state = stateFor(progress, id);
    if (isOwned(state) || vehicle.reserve) continue;
    const researched = Math.min(Math.max(Number(state.researchedRp || 0), 0), vehicle.rp);
    const remaining = Math.max(vehicle.rp - researched, 0);
    rp += remaining;
    ge += remaining ? Math.ceil(remaining / rpPerGe) : 0;
    sl += applyDiscount(Number(vehicle.sl || 0), options.slDiscountPercent);
  }
  const primary = options.optimizeFor === "sl" ? sl
    : options.optimizeFor === "vehicles" ? newIds.length
      : options.optimizeFor === "rp" ? rp : ge;
  return [primary, ge, sl, newIds.sort().join("\u0000")];
}

function minimumRankAdditions(database, base, countryId, branchId, rank, requiredCount,
  progress, options, allowReqUnlock, index, rpPerGe) {
  const candidates = treeVehicles(database, countryId, branchId).filter(vehicle =>
    vehicle.rank === rank && !base.has(vehicle.id)
    && (options.includeHiddenLegacy || !vehicle.hiddenResearch)
    && (allowReqUnlock || !vehicle.reqUnlock));
  if (!candidates.length) return new Set();

  let sequence = 0;
  const heap = new MinHeap((a, b) => compareTuples(a.cost, b.cost) || a.sequence - b.sequence);
  heap.push({cost: candidateCost(new Set(), base, index, progress, options, rpPerGe), sequence: sequence += 1, ids: new Set()});
  const visited = new Set();
  let processed = 0;
  while (heap.size) {
    const state = heap.pop();
    const signature = [...state.ids].sort().join("\u0000");
    if (visited.has(signature)) continue;
    visited.add(signature);
    processed += 1;
    if (processed > 75000) throw new Error("Rangoptimierung hat das Sicherheitslimit erreicht.");
    const combined = new Set([...base, ...state.ids]);
    if (countRank(combined, rank, index) >= requiredCount) return state.ids;

    for (const candidate of candidates) {
      if (combined.has(candidate.id)) continue;
      const addition = closure(database, candidate.id).filter(id =>
        !base.has(id) && (options.includeHiddenLegacy || !index.get(id).hiddenResearch));
      const nextIds = new Set([...state.ids, ...addition]);
      const nextSignature = [...nextIds].sort().join("\u0000");
      if (visited.has(nextSignature)) continue;
      heap.push({
        cost: candidateCost(nextIds, base, index, progress, options, rpPerGe),
        sequence: sequence += 1,
        ids: nextIds,
      });
    }
  }
  return new Set();
}

export function calculate(database, input = {}) {
  validateDatabase(database);
  const index = vehicleIndex(database);
  const target = index.get(input.targetId);
  const start = input.startId ? index.get(input.startId) : null;
  if (!target) throw new Error("Bitte ein Ziel auswählen.");
  if (input.startId && !start) throw new Error(`Unbekanntes Fahrzeug: ${input.startId}`);
  if (start && (start.countryId !== target.countryId || start.branchId !== target.branchId)) {
    throw new Error("Start und Ziel müssen im selben Forschungsbaum liegen.");
  }

  const progressInput = input.progress || {};
  const progress = {
    ...progressInput,
    vehicles: {...(progressInput.vehicles || {})},
    ownedGe: progressInput.ownedGe ?? input.ownedGe,
    convertibleRp: progressInput.convertibleRp ?? input.convertibleRp,
  };
  if (input.partialRp != null && !progress.vehicles[target.id]) {
    progress.vehicles[target.id] = {researchedRp: Number(input.partialRp) || 0};
  }
  const options = {
    optimizeFor: input.optimizeFor || "ge",
    includeStartVehicle: Boolean(input.includeStartVehicle),
    includeHiddenLegacy: Boolean(input.includeHiddenLegacy),
    slDiscountPercent: Number(input.slDiscount || 0),
  };
  if (!new Set(["ge", "rp", "sl", "vehicles"]).has(options.optimizeFor)) {
    throw new Error("Unbekanntes Optimierungsziel.");
  }
  applyDiscount(0, options.slDiscountPercent);
  if (target.hiddenResearch && !options.includeHiddenLegacy) {
    throw new Error(`${target.name || target.id} ist ein ausgeblendetes Altbestandsfahrzeug.`);
  }
  const rpPerGe = Number(database.economy?.rpPerGE ?? 45);
  const owned = new Set();
  for (const [id, state] of Object.entries(progress.vehicles || {})) {
    const vehicle = index.get(id);
    if (vehicle && isOwned(state) && vehicle.countryId === target.countryId
      && vehicle.branchId === target.branchId) {
      for (const predecessor of closure(database, id)) owned.add(predecessor);
    }
  }
  if (start) {
    const startClosure = closure(database, start.id);
    for (const id of startClosure) {
      if (!options.includeStartVehicle || id !== start.id) owned.add(id);
    }
  }

  const required = new Set(closure(database, target.id).filter(id => !owned.has(id)));
  const reasons = new Map([...required].map(id => [id, "direct_path"]));
  for (const vehicle of treeVehicles(database, target.countryId, target.branchId)) {
    if (vehicle.reserve) owned.add(vehicle.id);
  }

  const rankRequirements = [];
  const firstRank = start ? start.rank : 1;
  for (let rank = firstRank; rank < target.rank; rank += 1) {
    const needed = Number(database.rankUnlock?.[target.countryId]?.[target.branchId]?.[String(rank)] || 0);
    if (needed <= 0) continue;
    const before = new Set([...owned, ...required]);
    const availableBefore = countRank(before, rank, index);
    let added = new Set();
    if (availableBefore < needed) {
      added = minimumRankAdditions(database, before, target.countryId, target.branchId, rank,
        needed, progress, options, Boolean(start) || Boolean(owned.size), index, rpPerGe);
      for (const id of added) {
        required.add(id);
        if (!reasons.has(id)) reasons.set(id, "rank_unlock");
      }
    }
    const availableAfter = countRank(new Set([...owned, ...required]), rank, index);
    if (availableAfter < needed) {
      throw new Error(`Rang ${rank + 1} kann nicht geöffnet werden: ${availableAfter}/${needed} Fahrzeuge.`);
    }
    rankRequirements.push({
      rank, required: needed, availableBefore, availableAfter,
      addedVehicleIds: [...added].filter(id => index.get(id).rank === rank)
        .sort((a, b) => compareTuples(sortKey(index.get(a)), sortKey(index.get(b)))),
    });
  }
  if (start && options.includeStartVehicle) {
    required.add(start.id);
    reasons.set(start.id, "start_vehicle");
  }

  const warnings = [];
  const lines = [...required]
    .sort((a, b) => compareTuples(sortKey(index.get(a)), sortKey(index.get(b))))
    .map(id => {
      const vehicle = index.get(id);
      const state = stateFor(progress, id);
      const alreadyOwned = isOwned(state) || owned.has(id);
      const researched = Math.min(Math.max(Number(state.researchedRp || 0), 0), vehicle.rp);
      const remainingRp = alreadyOwned ? 0 : Math.max(vehicle.rp - researched, 0);
      if (vehicle.reqUnlock) warnings.push(`${vehicle.name}: zusätzliche Freischaltung ${vehicle.reqUnlock}`);
      if (vehicle.hiddenResearch) warnings.push(`${vehicle.name}: Altbestandsfahrzeug`);
      return {
        id, name: vehicle.name || id, reason: reasons.get(id) || "rank_unlock",
        totalRp: vehicle.rp, researchedRp: researched, remainingRp,
        ge: remainingRp ? Math.ceil(remainingRp / rpPerGe) : 0,
        sl: alreadyOwned ? 0 : applyDiscount(Number(vehicle.sl || 0), options.slDiscountPercent),
        alreadyOwned,
      };
    });
  const totalRp = lines.reduce((sum, line) => sum + line.remainingRp, 0);
  const totalGeBeforeOwned = lines.reduce((sum, line) => sum + line.ge, 0);
  const ownedGe = Math.max(Number(progress.ownedGe ?? input.ownedGe ?? 0), 0);
  const convertibleRp = progress.convertibleRp ?? input.convertibleRp;
  return {
    startVehicleId: start?.id || null,
    targetVehicleId: target.id,
    lines,
    rankRequirements,
    requiredVehicleIds: lines.map(line => line.id),
    totalRp,
    totalGeBeforeOwned,
    totalGe: Math.max(totalGeBeforeOwned - ownedGe, 0),
    totalSl: lines.reduce((sum, line) => sum + line.sl, 0),
    convertibleRpShortfall: convertibleRp == null ? 0 : Math.max(totalRp - Number(convertibleRp), 0),
    warnings: [...new Set(warnings)],
  };
}
